### ISSUE TITLE
Enable TokenRequest API?

### ISSUE THREAD

**Issue by zer0def** [schu/kubedee#59]
Some software packages started leveraging [service account token volume projection](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection), most notably - Istio. Would it be much of a bother to enable this feature as part of Kubedee's cluster deployment?

The addition of the below snippet to `kube-apiserver.service` would hopefully achieve the result, though YMMV:
```
  --service-account-api-audiences=api,vault,factors \\
  --service-account-issuer=api \\
  --service-account-signing-key-file=/etc/kubernetes/ca-key.pem \\
```

Helmfile installing Istio for replication purposes via a quick `helmfile sync`:
```
{{- $helmTimeout := default 600 (env "HELM_TIMEOUT") -}}
{{- $networkNamespace := default "network" (env "NETWORK_NAMESPACE") -}}
{{- $istioInitReleaseName := default "istio-init" (env "ISTIO_INIT_HELM_RELEASE_NAME") -}}
{{- $istioReleaseName := default "istio" (env "ISTIO_HELM_RELEASE_NAME") -}}
{{- $istioVersion := default "1.5.1" (env "ISTIO_VERSION") -}}
{{- $istioMajor := slice (splitList "." $istioVersion) 0 2 | join "." -}} 

repositories:
- name: istio{{ $istioMajor }}
  url: https://storage.googleapis.com/istio-release/releases/{{ $istioVersion }}/charts

helmDefaults:
  wait: true
  timeout: {{ $helmTimeout }}
  tillerless: false

releases:
- name: {{ $istioInitReleaseName }}
  namespace: {{ $networkNamespace }}
  chart: istio{{ $istioMajor }}/istio-init
  values:
  - certmanager:
      enabled: true
  hooks:
  - events: ["postsync"]
    command: "/bin/sh"
    args: ["-xec", "kubectl -n {{ $networkNamespace }} wait --for condition=complete --timeout {{ $helmTimeout }}s job --all && sleep 5"]
- name: {{ $istioReleaseName }}
  namespace: {{ $networkNamespace }}
  chart: istio{{ $istioMajor }}/istio
  needs:
  - {{ $networkNamespace }}/{{ $istioInitReleaseName }}
  values:
  - global:
      tag: {{ $istioVersion }}
      sds:
        enabled: true
    gateways:
      istio-ingressgateway:
        type: NodePort
        sds:
          enabled: true
```
The symptom of it not being there would be some of the Istio pods being stuck in `ContainerCreating`, because the feature is not available.

As of now, `istio-citadel` keeps crashing, so I'm probably supplying a bad trust chain in the first snippet.

**Comment by schu** [comment]
Hi, in general I'm happy to enable service account volume projection - can you send a pull request?

```
  --service-account-api-audiences=api,vault,factors \\
  --service-account-issuer=api \\
  --service-account-signing-key-file=/etc/kubernetes/ca-key.pem \\
```

I'm not sure I understand the configuration details for `--service-account-api-audiences` and `--service-account-issuer`:

Does the `--service-account-api-audiences` flag exist? I see it's mentioned in the [documentation for service accounts](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection) but neither in https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/ nor in `kube-apiserver -h` (v1.18.2). There's `--api-audiences` though - mistake in the docs or am I missing something?

I understand the `--service-account-api-audiences` value is a list of allowed (JWT) audiences, but I don't understand how the audiences are used in the context of projected service accounts and the kube-apiserver. Do you have a link at hand where this is explained? What are `vault` and `factor` audiences?

`--service-account-issuer`: IIUC this must be an https URL. Is this not true in the typical Kubernetes context?

Unfortunately the Kubernetes documentation is rather vague and cryptic here, IMHO.

Thanks.

**Comment by zer0def** [comment]
That's just a snippet I managed to pull from elsewhere and I'll submit a PR once I have something better fleshed out than what I've provided so far.

**Comment by schu** [comment]
> I understand the --service-account-api-audiences value is a list of allowed (JWT) audiences, but I don't understand how the audiences are used in the context of projected service accounts and the kube-apiserver. Do you have a link at hand where this is explained? What are vault and factor audiences?

After a bit more reading: as far as I understand,

* audiences are used to optionally narrow down the scope / purpose of a JWT token for third party applications, but not used by kube-apiserver (or rather the service account token authenticator) beyond checking that `aud` of a presented JWT is in the configured list.

* kube-apiserver allows requests, as long as the JWT is a) valid and b) the `aud` can be found in the configured list. There doesn't seem to be a special `aud` to define if a token can be used to speak to the Kubernetes API itself as opposed to only a third-party service with "Kubernetes authentication" (e.g. [vault configured to accept Kubernetes service account tokens for authentication](https://www.vaultproject.io/docs/auth/kubernetes)).?

* one could configure for example [vault](https://www.vaultproject.io/api/auth/kubernetes#audience) to only accept tokens that have a particular `aud: vault` (`--audience "vault"`). I assume the author of the snippet had that use-case in mind. So another pod that also has a valid JWT token from the same Kubernetes API server but with `aud: fabric` could authenticate against e.g the Kubernetes API server but would be rejected by the Vault instance (because the Vault instance would see: this is a valid token, but the audience field (i.e. intended purpose) is not right).

So I would suggest we start with the `--service-account-api-audiences` needed for your (Istio's) use case and add others later, when needed.

**Comment by schu** [comment]
> There doesn't seem to be a special aud to define if a token can be used to speak to the Kubernetes API itself as opposed to only a third-party service with "Kubernetes authentication"

... which I guess makes sense, because in Kubernetes land this is typically done via RBAC, i.e. to have a token which effectively can only be used for authenticating with third party APIs, don't assign any roles / role bindings to it.

**Comment by zer0def** [comment]
Based on [a comment in an Istio issue related to this](https://github.com/istio/istio/issues/17378#issuecomment-535304521), it looks like only `--service-account-issuer=kubernetes.default.svc --service-account-signing-key-file=/etc/kubernetes/ca-key.pem` need to be added. Submitting a PR shortly.

**Comment by zer0def** [comment]
Since this API has been made generally available in K8S 1.20, and by that extent is enabled by default, this issue isn't optional going into the future anymore, so it can be closed.
